zipline/currency.py

```diff
@@ -0,0 +1,70 @@
+from functools import partial, total_ordering
+
+from iso4217 import Currency as ISO4217Currency
+
+import numpy as np
+
+_ALL_CURRENCIES = {}
+
+
+def strs_to_sids(strs, category_num):
+    """TODO: Improve this.
+    """
+    out = np.full(len(strs), category_num << 50, dtype='i8')
+    casted_buffer = np.ndarray(
+        shape=out.shape,
+        dtype='S6',
+        buffer=out,
+        strides=out.strides,
+    )
+    casted_buffer[:] = np.array(strs, dtype='S6')
+    return out
+
+
+def str_to_sid(str_, category_num):
+    return strs_to_sids([str_], category_num)[0]
+
+
+iso_currency_to_sid = partial(str_to_sid, category_num=3)
+
+
+@total_ordering
+class Currency(object):
+    """A currency identifier, as defined by ISO-4217.
+
+    Parameters
+    ----------
+    code : str
+        ISO-4217 code for the currency.
+    """
+    def __new__(cls, code):
+        try:
+            return _ALL_CURRENCIES[code]
+        except KeyError:
+            iso_currency = ISO4217Currency(code)
```

Comment:
So, at the moment, this type isn't actually being used very much. I had originally added it as part of a larger branch that included a "currency domain" that allowed you to run pipelines that output currency data. On that branch, this type took took the place of the `Equity` object as part of the index of pipeline outputs.

On this branch, I believe the only place that `Currency` actually gets constructed is via `BoundColumn.fx`, which accepts either a `Currency` or a string and converts the string into a currency. For that functionality we could honestly probably get by with just using the string code, but think we're going to want the more complete object eventually, so it seemed cleaner to just use it from the start.

As far as the `bytes -> Currency` conversion goes, my expectation is that once we support currency domain outputs (and I think we'll want to do so relatively soon), the conversion will happen in roughly the same places where we currently convert from sid to Equity. For performance reasons, we tend to try to stay with the "low-level" representation (i.e., arrays of int64 or S3) inside the pipeline machinery, only converting to the object represnetation at the very end of the pipeline run. Concretely, in pipeline, this currently happens [here](https://github.com/quantopian/zipline/blob/master/zipline/pipeline/engine.py#L788). I expect we'll want to do something similar with currencies.

All of which is to say that I don't think there will be much value in being lenient w/r/t the input types here. There will be relatively few construction sites for Currency objects, and they'll have well-defined inputs and outputs. I'd rather have this type be strict about what it expects and have clients be responsible for the conversion. We can always make this more lenient in the future if we end up having a mix of call sites with different convenient input types, but it's relatively hard to go the other way (once you have an object that "helps" callers, it's harder to go back and fix all the callers that are doing the wrong thing).

Reply:
> Just wondering because of all of the bytes that are floating around for currency values in non-zipline places.

Also worth noting, I did end up having to add a conversion in: the `InMemoryExchangeRatesReader`: https://github.com/quantopian/zipline/pull/2586/commits/8d7cc53a02e7480166cb6540a5d713108869785f.